1.0.0 Features

dilyand opened the pull request.

istreeter commented on common/src/main/scala/com/snowplowanalytics/snowplow/storage/bigquery/common/config/CliConfig.scala:
```diff
@@ -0,0 +1,96 @@
+)
+
+object CliConfig {
+  sealed trait Environment {
+    val config: Config
+    val resolverJson: Json
+    val projectId: String
+
+    def getFullSubName(sub: String): String     = s"projects/$projectId/subscriptions/$sub"
+    def getFullTopicName(topic: String): String = s"projects/$projectId/topics/$topic"
+  }
```
Here's an interesting thing.  The code base still compiles if you make this a regular trait, not a sealed trait, and if you remove config and resolverJson as trait fields:

```scala
  trait Environment {
    val projectId: String

    def getFullSubName(sub: String): String     = s"projects/$projectId/subscriptions/$sub"
    def getFullTopicName(topic: String): String = s"projects/$projectId/topics/$topic"
  }
```

It might be a sign that sealed trait is not the tool you want for this job.  You are using this trait more as a mixin to add the `getFullSubName` and `getFullTopicName` methods.


You don't have to copy this, but here is a completely alternative approach:

```scala
case class Environment[A](config: A, resolverJson: Json, projectId: String) {
    def getFullSubName(sub: String): String     = s"projects/$projectId/subscriptions/$sub"
    def getFullTopicName(topic: String): String = s"projects/$projectId/topics/$topic"
}

type LoaderEnvironment = Environment[Config.Loader]
type MutatorEnvironment = Environment[Config.Mutator]
type RepeqterEnvironment = Environment[Config.Repeater]
```

dilyand replied:
Hey Ian, that is interesting indeed. I did try at some point making it a parameterised trait but it didn't work out and maybe the reason was that I should have also switched it to be a case class.

Other than succinctness, are there any additional benefits to this alternative approach? I believe my initial attempt at parameterisation was meant to cut down on the number of implicit codecs that needed to be derived, but later the Environment trait got moved out of the model and didn't really need codecs.

istreeter commented on loader/src/main/scala/com/snowplowanalytics/snowplow/storage/bigquery/loader/LoaderCli.scala:
```diff
@@ -10,26 +10,28 @@
+      c <- decodeBase64Hocon(args("config"))
       r <- decodeBase64Json(args("resolver"))
-      e <- transform[IO](EnvironmentConfig(r, c)).value.unsafeRunSync()
+      _ = validateResolverJson[IO](r).value.unsafeRunSync()
```
Have you checked whether this correctly rejects an invalid resolver json?  Because I think this line should be `<-` instead of `=` i.e.

```
_ <- validateResolverJson[IO](r).value.unsafeRunSync()
```

(you will need to leftMap the either you to a Throwable instead of a String)

Can you get it to work with the `[Id]` effect type instead of `[IO]`?  Then you don't need the ugly unsafeRunSync.

dilyand replied:
These are good points and I'll pay attention to them but I'm also thinking now that this check should probably be moved out of the modules and into the `decodeBase64Json` function in the main `Cli` file.

istreeter commented on mutator/src/main/scala/com/snowplowanalytics/snowplow/storage/bigquery/mutator/TypeReceiver.scala:
```diff
@@ -102,9 +102,9 @@ object TypeReceiver {
-      val subscription = ProjectSubscriptionName.of(config.projectId, config.typesSubscription)
+      val subscription = ProjectSubscriptionName.of(env.projectId, env.config.input.subscription)
       val subscriber   = Subscriber.newBuilder(subscription, listener).setHeaderProvider(UserAgent).build()
       subscriber.startAsync().awaitRunning(10L, TimeUnit.SECONDS)
```
I know nothing about this subscriber or what it does 😉 but if it is awaiting something something for 10 seconds, then I would expect to see a cats effect `Blocker` being used.  I know you didn't write this line of code so it's fine if you want to leave it.

chuwy replied:
I think it uses own threadpool underneath, so we probably had to use `IO.async`

istreeter commented on project/BuildSettings.scala:
```diff
@@ -127,7 +121,7 @@ object BuildSettings {
     Universal / sourceDirectory := new java.io.File((LocalRootProject / baseDirectory).value, "docker"),
-    Docker/ maintainer := "Snowplow Analytics Ltd. <support@snowplowanalytics.com>",
+    Docker    / maintainer := "Snowplow Analytics Ltd. <support@snowplowanalytics.com>",
     dockerBaseImage := "snowplow/base-debian:0.2.2",
```
I would like to see us move away from `snowplow/base-debian` and instead use `adoptopenjdk:11-jre-hotspot-focal` as a base image.  Again, I know this is completely unrelated to the current PR.  If you're interested you can take a look at where I did this [for the snowplow postgres loader](https://github.com/snowplow-incubator/snowplow-postgres-loader/blob/aa9a2a3da08f2de6f32eb2cdbdb4b430f3cd5497/project/BuildSettings.scala#L50)

dilyand replied:
https://github.com/snowplow-incubator/snowplow-bigquery-loader/issues/188

istreeter commented on repeater/src/main/scala/com.snowplowanalytics.snowplow.storage.bigquery.repeater/Repeater.scala:
```diff
@@ -43,36 +40,31 @@ object Repeater extends SafeIOApp {
           case Left(NonFatal(e)) =>
             System.out.println(e.toString)
             IO.raiseError(e) >> IO.pure(ExitCode.Error)
+          case Left(e) =>
```
This smells wrong.  You have already handled `Left(NonFatal(e))` on a line above.  So this line would only match a **fatal** exception.  Generally we don't deliberately catch fatal exceptions - they are things like `VirtualMachineError` or `ThreadDeath`.  I haven't checked where this exception is coming from, but you will hopefully find that it is already guaranteed to be non fatal or else it would not have been caught and wrapped into an Either.

istreeter commented on repeater/src/main/scala/com.snowplowanalytics.snowplow.storage.bigquery.repeater/Resources.scala:
```diff
@@ -26,40 +27,43 @@ import io.chrisdavenport.log4cats.Logger
   val jobStartTime: Instant
 ) {
-  def logInserted: F[Unit] =
+  def logInserted(): F[Unit] =
```
I'm not going to argue about this one.... but why the opinionated change from no parentheses to parentheses?

dilyand replied:
That was an IDE suggestion, I just assumed it was about some style convention to clearly flag `Unit` return types.

dilyand replied:
Apparently, it's in the Programming in Scala book:

> In principle it's possible to leave out all empty parentheses in Scala function calls. However, it is recommended to still write the empty parentheses when the invoked method represents more than a property of its receiver object. For instance, empty parentheses are appropriate if the method performs I/O, or writes reassignable variables (vars), or reads vars other than the receiver's fields, either directly or indirectly by using mutable objects. That way, the parameter list acts as a visual clue that some interesting computation is triggered by the call. For instance:

```
  "hello".length  // no () because no side-effect
  println()       // better to not drop the ()
```

> To summarize, it is encouraged style in Scala to define methods that take no parameters and have no side effects as parameterless methods, i.e., leaving off the empty parentheses. On the other hand, you should never define a method that has side-effects without parentheses, because then invocations of that method would look like a field selection. So your clients might be surprised to see the side effects. Similarly, whenever you invoke a function that has side effects, be sure to include the empty parentheses when you write the invocation. Another way to think about this is if the function you're calling performs an operation, use the parentheses, but if it merely provides access to a property, leave the parentheses off.

istreeter replied:
True that it is good practise to use `()` when there is a side effect and omit them when there is no side effect.

But.....

Things get more complicated when the side effect is suspended in `F`.  Here, the `F[Unit]` has no side effects, until the "end of the world"™ when we call `program.unsafeRunSync()` -- and notice the very deliberate parentheses on that function!

I do not have a strong opinion on this, and I have seen it done both ways for suspended effects.  I don't know if there is an agreed best practise for this situation, and I have never seen it discussed anywhere.

dilyand replied:
What you say makes sense, and I also have no strong feelings about it. I ended up keeping the `()` because if the IDE is suggesting it, it will be an ongoing annoyance if left unchanged.

chuwy replied:
I think it might be some IDE bug if it doesn't recognize `F` - I'm in relatively strong opposition to `()`

dilyand replied:
I reverted those changes.

This ended up being the mostly hotly discussed change in the PR; but would you just look at the lovely bike shed we have now. :)

istreeter commented on repeater/src/main/scala/com.snowplowanalytics.snowplow.storage.bigquery.repeater/Resources.scala:
```diff
@@ -71,92 +75,96 @@ class Resources[F[_]: Sync](
+      bucket <- Sync[F].fromEither(
+        validateBucket(env.config.output.deadLetters.bucket)
+          .toEither
+          .leftMap(e => throw new RuntimeException(e.toList.mkString("\n")))
```
The `throw` is wrong here.  You want to return the RuntimeException not throw it.

istreeter reviewed (commented):
This is a good PR, and they are nice improvements you made to the config parsing.  I have left a smattering of comments but there is nothing wrong with the general approach you have taken.

dilyand commented:
@istreeter I addressed your comments in separate commits to hopefully make it easier to re-review. If you are happy, please green tick this PR, so it can be merged into `develop`.

istreeter commented on repeater/src/main/scala/com.snowplowanalytics.snowplow.storage.bigquery.repeater/Repeater.scala:
```diff
@@ -62,10 +62,8 @@ object Repeater extends SafeIOApp {
             IO.raiseError(e) >> IO.pure(ExitCode.Error)
-          case Left(e) =>
-            IO(System.err.println(s"Snowplow BigQuery Repeater failed to start. ${e.toString}")) >> IO
-              .pure(ExitCode.Error)
```
I'm slightly confused now, because I would expect the compiler to complain about non-exhaustive pattern matching.  By matching just `Left(NonFatal(e))` it means nothing is matching `Left(e)` where e _is_ fatal.  I checked the compiler but there are no warnings.  To be clear, I don't want us to catch and handle fatal warnings.  But pattern matching in a `flatMap` is a different case, because the exception has already been caught and wrapped into an Either.

I don't think we need to investigate this further though.  If the compiler is happy then I am happy.

dilyand replied:
The way I understand it, we want to allow fatal errors to stop the program; and if `flatMap` sees an unmatched case it will throw and exception which will have the same effect.

chuwy commented on common/src/test/scala/com/snowplowanalytics/snowplow/storage/bigquery/common/CliConfigSpec.scala:
```diff
@@ -0,0 +1,41 @@
+      decodeBase64Hocon(SpecHelpers.validHoconB64) must beRight
+    }
+    "fail with invalid HOCON" in {
+      decodeBase64Hocon(SpecHelpers.invalidHoconB64) must beLeft
```
I recommend you to take the error message into account - sometimes parser produces very cryptic messages and we want to make sure there are human-readable strings.

chuwy commented on loader/src/main/scala/com/snowplowanalytics/snowplow/storage/bigquery/loader/metrics.scala:
```diff
@@ -10,8 +10,7 @@
  * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
  */
-package com.snowplowanalytics.snowplow.storage.bigquery
-package loader
```
👍  

I hate that namespacing approach.

chuwy commented on mutator/src/main/scala/com/snowplowanalytics/snowplow/storage/bigquery/mutator/MutatorCli.scala:
```diff
@@ -0,0 +1,63 @@
+  private val verbose: Opts[Boolean] = Opts.flag("verbose", "Provide debug output").orFalse
+
+  sealed trait MutatorCommand extends Product with Serializable
+  final case class CreateCommand(env: MutatorEnvironment) extends MutatorCommand
+  final case class ListenCommand(env: MutatorEnvironment, verbose: Boolean) extends MutatorCommand
+  final case class AddColumnCommand(env: MutatorEnvironment, schema: SchemaKey, property: ShredProperty)
+      extends MutatorCommand
```
```suggestion
  object MutatorCommand {
    final case class CreateCommand(env: MutatorEnvironment) extends MutatorCommand
    final case class ListenCommand(env: MutatorEnvironment, verbose: Boolean) extends MutatorCommand
    final case class AddColumnCommand(env: MutatorEnvironment, schema: SchemaKey, property: ShredProperty)
        extends MutatorCommand
  }
```

chuwy reviewed (approved):
Looks great!

I'd just double-check that HOCON parser produce sensible error messages because pureconfig has a bad track there.
